### PR TITLE
Support custom implementations of mobile units

### PR DESCRIPTION
**Pull request by rossipaolo**
The same as  #1598, this time for enemies. 

[ExampleMobileUnit.zip](https://github.com/Interkarma/daggerfall-unity/files/5860569/ExampleMobileUnit.zip)

### CODE REVIEW

File: Assets/Scripts/Internal/DaggerfallMobileBillboard.cs [hunk 1]
```diff
@@ -0,0 +1,734 @@
...
+        }
+
+        /// <summary>
+        /// Sets a new enemy state and restarts frame counter.
+        /// Certain states are one-shot only (such as attack and hurt) and return to idle when completed.
+        /// Continuous states (such as move) will keep looping until changed.
+        /// </summary>
+        /// <param name="state">New state.</param>
+        /// <returns>Frames per second of new state.</returns>
+        public override float ChangeEnemyState(MobileStates state)
```
Comment (TheLacus) [inline Assets/Scripts/Internal/DaggerfallMobileBillboard.cs]: Return value is never used and specific to billboard animation. Can i remove it?

File: Assets/Scripts/Internal/DaggerfallMobileBillboard.cs [hunk 2]
```diff
@@ -0,0 +1,734 @@
...
+
+        public sealed override MobileStates EnemyState
+        {
+            get { return summary.EnemyState; }
+        }
+
+        public sealed override int AnimStateRecord
+        {
+            get { return summary.AnimStateRecord; }
+        }
```
Comment (TheLacus) [inline Assets/Scripts/Internal/DaggerfallMobileBillboard.cs]: Used only here:

```cs
int damage = FormulaHelper.CalculateAttackDamage(playerEntity, enemyEntity, entityMobileUnit.AnimStateRecord, animTime, strikingWeapon);
```

I understand is used to detect backstabbing:

```cs
// If enemy is facing away from player
if (enemyAnimStateRecord % 5 > 2)
{
    player.TallySkill(DFCareer.Skills.Backstabbing, 1);
    return player.Skills.GetLiveSkillValue(DFCareer.Skills.Backstabbing);
}
```

Can i replace it with `bool IsBackFacing`?

Comment (ajrb) [reply Assets/Scripts/Internal/DaggerfallMobileBillboard.cs]: That would make a much better interface IMO. so +1 from me

**Comment by ajrb** [comment]
Generally seems fine, although this is not code I am familiar with so I may well miss something.

My understanding is that this allows the visuals for enemies to be changed into models, is that correct? Pretty sure the sprites can already be replaced.

**Comment by TheLacus** [comment]
Yes, it allows to provide a custom implementation of enemy graphics. For example MasonFace is currently [trying to use an opensource framework for npcs](https://forums.dfworkshop.net/viewtopic.php?f=23&t=3498); i think [he plans to do monsters next](https://forums.dfworkshop.net/viewtopic.php?p=40859#p40859).

**Comment by TheLacus** [comment]
df_enemy

**Comment by Interkarma** [comment]
I love where this is going. Thank you for help with review. I'll take a look at this one more closely when I can, and once conflicts resolved.

File: Assets/Scripts/Internal/DaggerfallMobileBillboard.cs (hunk 1 above)
Comment (Interkarma) [reply Assets/Scripts/Internal/DaggerfallMobileBillboard.cs]: Apologies for not answering. If never used, happy for this return value to be removed.

**Comment by Interkarma** [comment]
Hey @TheLacus! :) My apologies for losing track of this PR. I see you're still working on this one. Just ping me when out of draft and ready for review. I don't have any issues with how you're implementing this so far. I'll do my best to test and help find any issues once you're happy with it. Then we can move into master finally. :)

**Comment by TheLacus** [comment]
Thank you!  jefetienne also proposed to help testing, so hopefully we'll get this feature available to mod authors soon.

**Comment by TheLacus** [comment]
I uploaded a simple mod that can help with testing. My understanding is that combat behaviour (`DoMeleeDamage` and `ShootArrow`) is strictly linked with billboard animation frames, so mod authors need to reimplement both asset handling and combat for specific enemies. I'm not sure if we can make it a bit easier to manage, but in any case i think is something that can be revised at a later time.

**Comment by Interkarma** [comment]
Thanks for all your hard work on this one @rossipaolo. I'm happy with all your changes and haven't been able to find any issues in base game. Really looking forward to watching this one evolve as modders start using this.

**Pull request merged**